docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+**C and Objective-C interoperability:** C++ interoperability will have to build off of the previous art of C and Objective-C interoperability. Starting with the name, C++ is a superset of C, which means C++ interoperability has a great starting place to build off of. C is a simple language without many high level constructs, so importing C APIs is, relatively, simple. On the other hand, Objective-C adds some high level constructs, such as reference types and a generics model. However, unlike C++, Objective-C has strong idioms; there are only so many patterns for doing some specific thing. Further, Objective-C was able to be updated to better interoperate with Swift, the generics model is a good example of this. So, while C and Objective-C interoperability is a good starting place, the model for importing these languages cannot be taken wholesale for C++ interoperability. 
+
+## Goals
+
+Safety is a top priority. The compiler should not import APIs that would be less safe in Swift than they are in C++ and interop should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators).
+
+C++ interoperability aims to provide a clear, well-defined mapping for APIs that can be imported, and to draw a clear boundary between these APIs and APIs that cannot be imported. A user should be able to read the C++ interoperability documentation and have a good idea of both how much of their C++ API will be able to be imported in Swift and what that API will look like once it is imported. 
+
+It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible. Imported APIs, especially containers, should not be implicitly converted to Swift types (in contrast to the automatic Objective-C bridging for NSArray and NSString). 
+
+It is a goal to preserve the semantics (including performance semantics) that an API was written with by understanding the API’s semantics and mapping it to a Swift idiom that represents the same thing. 
+
+Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 
+
+It is *not* a goal to import every C++ API into Swift. APIs that are imported by Swift should feel Swifty and idiomatic. There must be an easy way to add annotations so that APIs become importable, especially when updating APIs directly is not an option (because the source code is not available). 
```

Comment:
Without some idea of which kinds of APIs you intend not to import, this statement is perhaps needlessly alarming.  For example, I can live without the ability to use function-style macros from Swift, but I can't think of any other category of C++ APIs that obviously should be off the table, and you can make interop arbitrarily useless to me depending on where you draw that line.

Keep in mind that `vector<bool>` is a specialization, and if you're going to prevent me from using `vector<T>` in code that's generic on `T`, that would be pretty hard to swallow.

Reply:
I disagree that it's needlessly alarming. If the goal of the first cut of this interop is to import only a manageable subset of C++, saying so in the goals section seems reasonable. It has to be said at some point, otherwise we're back to the previous (very challenging) plan of trying to import everything in a v1 implementation, which is what led us to here in the first place.